tests/test_prolific.py

```diff
@@ -115,3 +226,291 @@ def test_can_add_to_available_place_count(subject):
 
     assert updated["total_available_places"] == initial_spaces + 1
     assert subject.delete_study(study_id=result["id"])
+
+
+def test_translate_workspace_name_exception(subject):
+    """_translate_workspace_name raises ProlificServiceNoSuchWorkspaceException."""
+
+    # Mock out self._req.
+    subject._req = MagicMock(return_value=WORKSPACES_API_RETURN_VALUE)
+
+    with pytest.raises(ProlificServiceNoSuchWorkspaceException):
+        subject._translate_workspace_name("does not exist")
+
+
+def test_translate_project_name_exception(subject):
+    """_translate_project_name raises ProlificServiceNoSuchProjectException."""
+
+    # Mock out self._req.
+    subject._req = MagicMock(return_value=PROJECTS_API_RETURN_VALUE)
+
+    with pytest.raises(ProlificServiceNoSuchProjectException):
+        subject._translate_project_name("unused", "does not exist")
+
+
+@pytest.mark.parametrize(
+    "test_input, expected",
+    [
+        # Translate a name into an id.
+        ("zippy the pinhead", "66b0f8e34632badef5c8d1db"),
+        # Translate an id into an id.
+        ("66b0f8e34632badef5c8d1db", "66b0f8e34632badef5c8d1db"),
+    ],
+)
+def test_translate_workspace_name(subject, test_input, expected):
+    """_translate_workspace_name returns a workspace id, when called with either a name or an id."""
+
+    # Mock out self._req.
+    subject._req = MagicMock(return_value=WORKSPACES_API_RETURN_VALUE)
+
+    assert subject._translate_workspace_name(test_input) == expected
+
+
+@pytest.mark.parametrize(
+    "test_input, expected",
+    [
+        # Translate a name into an id.
+        ("default ws project", "66b0f923fa279fd68ab7bd54"),
+        # Translate an id into an id.
+        ("66b0f923fa279fd68ab7bd54", "66b0f923fa279fd68ab7bd54"),
+    ],
+)
+def test_translate_project_name(subject, test_input, expected):
+    """_translate_project_name returns a project id, when called with either a name or an id.."""
+
+    # Mock out self._req.
+    subject._req = MagicMock(return_value=PROJECTS_API_RETURN_VALUE)
+
+    assert subject._translate_project_name("unused", test_input) == expected
+
+
+@pytest.mark.parametrize(
```

Comment:
I don't mind the parameterized tests when the params are simple (as in the cases above), but I confess it hurts my brain quite badly when they get complicated, because there's so much to track. Curious why you chose this over multiple tests?

I don't recommend changing anything now; I may just need so time to appreciate this approach.

Reply:
I agree there's a trade-off.

With separate tests, a bugfix or change must be made to multiple test functions.  And it's more code overall.

With one parameterized test, all of the test variants are in one test function, so you edit only one function But that one function + the decorator can be lengthy.

I prefer using parameterize because the "less code, everything-in-one-function" benefit outweighs the cost IMO.  FWIW, I have in the past used separate tests because the parameter decorator went on for literally multiple screens.  That was too much for me. :)

Reply:
I just thought of a way to reduce the decorator length.  stay tuned